net-nntp/sabnzbd/files/sabnzbd.initd

```diff
@@ -1,99 +1,71 @@
+
+	checkpath -q -d -o ${SABNZBD_USER}:${SABNZBD_GROUP} -m 0770 "$(dirname "${PIDFILE}")"
+
+	start-stop-daemon \
+		--quiet \
+		--start \
+		--user ${SABNZBD_USER} \
+		--group ${SABNZBD_GROUP} \
+		--pidfile "${PIDFILE}" \
+		--wait 1000 \
+		--exec /usr/bin/sabnzbd \
+		-- \
+		--config-file "${SABNZBD_CONFIGFILE}" \
+		--logging "${SABNZBD_LOGGING}" \
+		--daemon \
+		--pidfile "${PIDFILE}"
+
+	eend $?
 }
 
-get_addr() {
-    local host=$(get_var 'host')
-    local protocol='http'
+stop() {
+	local protocol="http"
+	local host="$(get_var "host")"
+	local port="$(get_var "port")"
 
-    [ "${host}" == "0.0.0.0" ] && host=localhost
-    [ "$(get_var 'enable_https')" -eq 1 ] && protocol='https'
+	if [ $(get_var "enable_https") -eq 1 ]; then
+		protocol="https"
+		port="$(get_var "https_port")"
+	fi
 
-    echo "${protocol}://${host}:$(get_port)"
-}
+	case "${host}" in
+		*:*) host="[${host}]" ;;
+	esac
 
-get_pidfile() {
-    echo "${RUNDIR}/sabnzbd-$(get_port).pid"
-}
+	local url="${protocol}://${host}:${port}/sabnzbd/api?mode=shutdown"
 
-start() {
-    ebegin "Starting SABnzbd"
-
-    checkpath -q -d -o ${SABNZBD_USER}:${SABNZBD_GROUP} -m 0770 "${RUNDIR}"
-
-    start-stop-daemon \
-        --quiet \
-        --start \
-        --user ${SABNZBD_USER} \
-        --group ${SABNZBD_GROUP} \
-        --name sabnzbd \
-        --background \
-        --pidfile "$(get_pidfile)" \
-        --exec /usr/bin/sabnzbd \
-        -- \
-        --config-file "${SABNZBD_CONFIGFILE}" \
-        --logging "${SABNZBD_LOGGING}" \
-        --daemon \
-        --pid "${RUNDIR}"
-
-    eend $?
-}
+	if [ $(get_var "disable_api_key") -eq 0 ]; then
+		url="${url}&apikey=$(get_var "api_key")"
+	fi
 
-start_pre() {
-    if [ "$RC_CMD" == "restart" ]; then
-        local pidfile=$(get_pidfile)
-        while [ -e ${pidfile} ]; do
-            sleep 1
-        done
-    fi
+	local signals="TERM/1/KILL/1"
 
-    return 0
-}
+	ebegin "Stopping SABnzbd"
 
-stop() {
-    local api_key=$(get_var 'api_key')
-    local addr=$(get_addr)
-    local rc=1
-
-    ebegin "Stopping SABnzbd @ ${addr}"
-    # This can only work if we have enabled the API
-    if [ -n "${api_key}" -a "$(get_var 'disable_api_key')" -ne 1 ]; then
-        local ret
-        einfo "Attempting web-based shutdown @ ${addr}"
-
-        # SABnzbd will return "ok" if shutdown is successful
-        ret=$(/usr/bin/curl -k -s "${addr}/sabnzbd/api?mode=shutdown&apikey=${api_key}")
-        [ "${ret}" == "ok" ] && rc=0
-    fi
-
-    if [ "${rc}" -ne 0 ]; then
-        einfo "Falling back to SIGTERM, this may not work if you restarted via the web interface"
-        start-stop-daemon \
-            --stop \
-            --pidfile $(get_pidfile) \
-            --retry SIGTERM/1/SIGKILL/5
-        rc=$?
-    fi
-
-    eend ${rc}
+	if [ "$(wget -q -t 1 -O - -T 10 "${url}")" = "ok" ]; then
+		signals="CONT/5/${signals}"
+	fi
+
+	start-stop-daemon \
+		--stop \
+		--pidfile "${PIDFILE}" \
+		--retry "${signals}"
```

Comment:
The `while` loop in the `start_pre()` function of the current version of this script might never end as it assumes the pidfile will disappear at some point.  This might not always be the case.  Thus, to enforce a timeout and yet not reinvent the wheel, how about using `start-stop-daemon`?

The code above first attempts an API-based shutdown.  If it succeeds, `start-stop-daemon` is asked to send SIGCONT to the process whose PID is present in the pidfile (which is essentially a noop for a running process) and wait for a maximum of 5 seconds for the process to actually exit.  If the process fails to exit cleanly before the timeout, `start-stop-daemon` will try to kill it with SIGTERM and finally, if everything else fails, with SIGKILL.  If the API-based shutdown fails in the first place, SIGTERM is sent immediately, followed by SIGKILL if needed.  If the pidfile is already removed by the time `start-stop-daemon` is called, the latter will return 0, which is also fine in this case.

Doing it this way voids to need to implement a `while` loop with timeout in shell and also fully delegates the generation of the return code to `start-stop-daemon`, without sacrificing any functionality of the original code.